pkg/integrations/prometheus/promql/parser.go

```diff
@@ -0,0 +1,564 @@
+		return []*structs.MetricsQueryRequest{}, "", []*structs.QueryArithmetic{}, err
+	}
+
+	pqlQuerytype := expr.Type()
+	var mQuery structs.MetricsQuery
+	mQuery.Aggregator = structs.Aggregation{}
+	selectors := extractSelectors(expr)
+	//go through labels
+	for _, lblEntry := range selectors {
+		for _, entry := range lblEntry {
+			if entry.Name != "__name__" {
+				tagFilter := &structs.TagsFilter{
+					TagKey:          entry.Name,
+					RawTagValue:     entry.Value,
+					HashTagValue:    xxhash.Sum64String(entry.Value),
+					TagOperator:     segutils.TagOperator(entry.Type),
+					LogicalOperator: segutils.And,
+				}
+				mQuery.TagsFilters = append(mQuery.TagsFilters, tagFilter)
+			} else {
+				mQuery.MetricName = entry.Value
+			}
+		}
+	}
+
+	timeRange := &dtu.MetricsTimeRange{
+		StartEpochSec: startTime,
+		EndEpochSec:   endTime,
+	}
+
+	mQuery.OrgId = myid
+	mQuery.PqlQueryType = pqlQuerytype
+
+	intervalSeconds, err := mresults.CalculateInterval(endTime - startTime)
+	if err != nil {
+		return []*structs.MetricsQueryRequest{}, "", []*structs.QueryArithmetic{}, err
+	}
+
+	metricQueryRequest := &structs.MetricsQueryRequest{
+		MetricsQuery: mQuery,
+		TimeRange:    *timeRange,
+	}
+
+	mQueryReqs := make([]*structs.MetricsQueryRequest, 0)
+	mQueryReqs = append(mQueryReqs, metricQueryRequest)
+
+	queryArithmetic := make([]*structs.QueryArithmetic, 0)
+	var exitFromInspect bool
+
+	parser.Inspect(expr, func(node parser.Node, path []parser.Node) error {
+		if !exitFromInspect {
+			mQueryReqs, queryArithmetic, exitFromInspect, err = parsePromQLExprNode(node, mQueryReqs, queryArithmetic, intervalSeconds)
+		}
+		return err
+	})
+
+	if len(mQueryReqs) == 0 {
+		return mQueryReqs, pqlQuerytype, queryArithmetic, nil
+	}
+
+	mQuery = mQueryReqs[0].MetricsQuery
+
+	tags := mQuery.TagsFilters
+	for idx, tag := range tags {
+		var hashedTagVal uint64
+		switch v := tag.RawTagValue.(type) {
+		case string:
+			hashedTagVal = xxhash.Sum64String(v)
+		case int64:
+			hashedTagVal = uint64(v)
+		case float64:
+			hashedTagVal = uint64(v)
+		case uint64:
+			hashedTagVal = v
+		default:
+			log.Errorf("ParseMetricsRequest: invalid tag value type")
+		}
+		tags[idx].HashTagValue = hashedTagVal
+	}
+
+	if mQuery.MQueryAggs == nil {
+		mQuery.MQueryAggs = &structs.MetricQueryAgg{
+			AggBlockType:    structs.AggregatorBlock,
+			AggregatorBlock: &structs.Aggregation{AggregatorFunction: segutils.Avg},
+		}
+		mQueryReqs[0].MetricsQuery = mQuery
+	} else {
+		// If the first Block in the MQueryAggs is not an aggregator block, then add an default aggregator block with avg function
+
+		if mQuery.MQueryAggs.AggBlockType != structs.AggregatorBlock {
+			mQuery.MQueryAggs = &structs.MetricQueryAgg{
+				AggBlockType:    structs.AggregatorBlock,
+				AggregatorBlock: &structs.Aggregation{AggregatorFunction: segutils.Avg},
+				Next:            mQuery.MQueryAggs,
+			}
+			mQueryReqs[0].MetricsQuery = mQuery
+		}
```

Comment:
I think you'd be writing over whatever was at position 0, right? Do you need to prepend this new item instead?

Reply:
Are you talking about line 157 or line 154. If it is line 154, then I am prepending it.
For line 157, yes I want to overwrite the first element and there will only be one element. This is applied in the case where each individual sub query in  a multi queries will be parsed once at a time.

Reply:
I'm talking about line 157, but I'm not understanding what you're saying.

Since we got here, `mQuery.MQueryAggs.AggBlockType != structs.AggregatorBlock`, so `mQuery.MQueryAggs.AggBlockType == FunctionBlock`. Lines 152-157 would write over this so you lose that `FunctionBlock` agg. Why is this the correct thing to do?

Reply:
At line 122: `mQuery = mQueryReqs[0].MetricsQuery`, I am taking the value of mQuery and it is a value type, so if I change here, it won't reflect in the `mQueryReqs` array.
So I am reassigning it. 

Also, the `FunctionBlock` won't be lost it will be in the next. We need the first block to be an `AggregatorBlock`, So when the first block is not an `AggregatorBlock`, we need to add the default aggregation Block which is what I was doing in the Lines 152 - 154 and at Line 155 which `Next`, I am assigning the current `mQueryAggs` by doing this `Next: mQuery.MQueryAggs`. 

So the first block is Aggregator block and the second will have the Function Block.

Reply:
Ok, makes sense